File: apps/api/src/routes/cron.ts
import { Hono } from 'hono'
import { cronService } from '../services/cron.service.js'

const app = new Hono()

app.get('/cron', async (c) => {
  const workspaceId = c.req.query('workspaceId') || undefined
  const sessionId = c.req.query('sessionId') || undefined
  const includeGlobal = c.req.query('includeGlobal')
  const includeWorkspace = c.req.query('includeWorkspace')
  const includeConversation = c.req.query('includeConversation')

  const jobs = await cronService.list({
    workspaceId,
    sessionId,
    includeGlobal: includeGlobal === undefined ? undefined : includeGlobal === 'true',
    includeWorkspace: includeWorkspace === undefined ? undefined : includeWorkspace === 'true',
    includeConversation:
      includeConversation === undefined ? undefined : includeConversation === 'true',
  })
  return c.json({ success: true, data: jobs })
})

app.post('/cron', async (c) => {
  const body = await c.req.json()
  const job = await cronService.create(body)
  return c.json({ success: true, data: job }, 201)
})

app.patch('/cron/:id', async (c) => {
  const id = c.req.param('id')
  const body = await c.req.json()
  const job = await cronService.update(id, body)
  return c.json({ success: true, data: job })
})

app.delete('/cron/:id', async (c) => {
  const id = c.req.param('id')
  try {
    await cronService.delete(id)
    return c.json({ success: true })
  } catch (err) {
    if (err instanceof Error && err.message.includes('built-in')) {
      return c.json({ success: false, message: err.message }, 403)
    }
    throw err
  }
})

app.patch('/cron/:id/toggle', async (c) => {
  const id = c.req.param('id')
  const { enabled } = await c.req.json()
  const job = await cronService.toggleEnabled(id, enabled)
  return c.json({ success: true, data: job })
})

app.post('/cron/:id/trigger', async (c) => {
  const id = c.req.param('id')
  await cronService.triggerNow(id)
  return c.json({ success: true })
})

export default app

File: backend/src/clawbuddy/routers/cron.py
"""Cron router — manage scheduled cron jobs.

Replaces: apps/api/src/routes/cron.ts
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from clawbuddy.db.session import get_db
from clawbuddy.lib.responses import fail, ok
from clawbuddy.services.cron import cron_service

router = APIRouter(tags=["cron"])


@router.get("/admin/cron")
async def list_cron_jobs(db: AsyncSession = Depends(get_db)):
    """List all cron jobs."""
    jobs = await cron_service.list_jobs(db)
    return ok([
        {
            "id": j.id,
            "name": j.name,
            "description": j.description,
            "schedule": j.schedule,
            "type": j.type,
            "handler": j.handler,
            "prompt": j.prompt,
            "workspaceId": j.workspace_id,
            "sessionId": j.session_id,
            "enabled": j.enabled,
            "builtin": j.builtin,
            "lastRunAt": j.last_run_at.isoformat() if j.last_run_at else None,
            "lastRunStatus": j.last_run_status,
            "lastRunError": j.last_run_error,
            "createdAt": j.created_at.isoformat() if j.created_at else None,
        }
        for j in jobs
    ])


@router.post("/admin/cron", status_code=201)
async def create_cron_job(
    body: dict[str, Any],
    db: AsyncSession = Depends(get_db),
):
    """Create a new cron job."""
    try:
        job = await cron_service.create(
            db,
            name=body.get("name", ""),
            schedule=body.get("schedule", ""),
            description=body.get("description"),
            job_type=body.get("type", "agent"),
            handler=body.get("handler"),
            prompt=body.get("prompt"),
            workspace_id=body.get("workspaceId"),
            session_id=body.get("sessionId"),
            enabled=body.get("enabled", True),
        )
        return ok({
            "id": job.id,
            "name": job.name,
            "schedule": job.schedule,
            "enabled": job.enabled,
        })
    except ValueError as exc:
        return fail(str(exc), status_code=400)


@router.patch("/admin/cron/{job_id}")
async def update_cron_job(
    job_id: str,
    body: dict[str, Any],
    db: AsyncSession = Depends(get_db),
):
    """Update a cron job."""
    try:
        job = await cron_service.update(db, job_id, **body)
        return ok({
            "id": job.id,
            "name": job.name,
            "schedule": job.schedule,
            "enabled": job.enabled,
        })
    except ValueError as exc:
        return fail(str(exc), status_code=400)


@router.delete("/admin/cron/{job_id}")
async def delete_cron_job(
    job_id: str,
    db: AsyncSession = Depends(get_db),
):
    """Delete a cron job (403 if builtin)."""
    try:
        await cron_service.delete_job(db, job_id)
        return ok(None)
    except ValueError as exc:
        if "built-in" in str(exc):
            return fail(str(exc), status_code=403)
        return fail(str(exc), status_code=400)


@router.patch("/admin/cron/{job_id}/toggle")
async def toggle_cron_job(
    job_id: str,
    body: dict[str, Any],
    db: AsyncSession = Depends(get_db),
):
    """Toggle a cron job enabled/disabled."""
    enabled = body.get("enabled", True)
    job = await cron_service.toggle_enabled(db, job_id, enabled)
    return ok({
        "id": job.id,
        "name": job.name,
        "enabled": job.enabled,
    })


@router.post("/admin/cron/{job_id}/trigger")
async def trigger_cron_job(
    job_id: str,
    db: AsyncSession = Depends(get_db),
):
    """Trigger a cron job for immediate execution."""
    await cron_service.trigger_now(db, job_id)
    return ok(None)
